File: prototype/forward_mode.cpp
#include "forward_mode.h"

#include <cmath>
#include <functional>
#include <iomanip>
#include <iostream>
#include <vector>

namespace ad {

Sym::Sym(double _val, double _dot) : m_val(_val), m_dot(_dot) {}

Sym::Sym(double _val) : m_val(_val) {}

auto Sym::value() const noexcept -> double { return m_val; }
auto Sym::dot() const noexcept -> double { return m_dot; }

auto operator+(const Sym &lhs, const Sym &rhs) -> Sym {
  return {lhs.value() + rhs.value(), lhs.dot() + rhs.dot()};
}

auto operator-(const Sym &lhs, const Sym &rhs) -> Sym {
  return {lhs.value() - rhs.value(), lhs.dot() - rhs.dot()};
}

auto operator*(const Sym &lhs, const Sym &rhs) -> Sym {
  return {lhs.value() * rhs.value(),
          lhs.dot() * rhs.value() + rhs.dot() * lhs.value()};
}

auto operator/(const Sym &lhs, const Sym &rhs) -> Sym {
  const double df = (lhs.dot() * rhs.value() - lhs.value() * rhs.dot()) /
                    std::pow(rhs.value(), 2);
  return {lhs.value() / rhs.value(), df};
}
// has problems to address df.dexp and df.dbase case
auto pow(const Sym &base, const Sym &exp) -> Sym {
  const double df_base =
      std::log(base.value()) * std::pow(base.value(), exp.value()) * exp.dot();
  return {std::pow(base.value(), exp.value()), df_base};
}

auto pow(const Sym &base, double exp) -> Sym {
  const double df_base = base.dot() * std::pow(base.value(), exp - 1);
  return {std::pow(base.value(), exp), df_base};
}

auto exp(const Sym &rhs) noexcept -> Sym {
  const double value = std::exp(rhs.value());
  const double df = value * rhs.dot();
  return {value, df};
}

auto ln(const Sym &rhs) noexcept -> Sym {
  const double value = std::log(rhs.value());
  const double df = (1.0f / rhs.value()) * rhs.dot();
  return {value, df};
}

auto sin(const Sym &rhs) noexcept -> Sym {
  const double value = std::sin(rhs.value());
  const double df = std::cos(rhs.value()) * rhs.dot();
  return {value, df};
}

auto cos(const Sym &rhs) noexcept -> Sym {
  const double value = std::cos(rhs.value());
  const double df = -std::sin(rhs.value()) * rhs.dot();
  return {value, df};
}

auto tan(const Sym &rhs) noexcept -> Sym {
  const double value = std::tan(rhs.value());
  const double df = std::pow(1.0f / std::cos(rhs.value()), 2) * rhs.dot();
  return {value, df};
}

auto cot(const Sym &rhs) noexcept -> Sym {
  const double value = 1.0f / std::tan(rhs.value());
  const double df = -std::pow(1.0f / std::sin(rhs.value()), 2) * rhs.dot();
  return {value, df};
}

auto sec(const Sym &rhs) noexcept -> Sym {
  const double value = 1.0f / std::cos(rhs.value());
  const double df = value * std::tan(rhs.value()) * rhs.dot();
  return {value, df};
}

auto csc(const Sym &rhs) noexcept -> Sym {
  const double value = 1.0f / std::sin(rhs.value());
  const double df = -value * (1.0f / std::tan(rhs.value())) * rhs.dot();
  return {value, df};
}

auto sinh(const Sym &rhs) noexcept -> Sym {
  const double value = std::sinh(rhs.value());
  const double df = std::cosh(rhs.value()) * rhs.dot();
  return {value, df};
}

auto cosh(const Sym &rhs) noexcept -> Sym {
  const double value = std::cosh(rhs.value());
  const double df = -std::sinh(rhs.value()) * rhs.dot();
  return {value, df};
}

auto tanh(const Sym &rhs) noexcept -> Sym {
  const double value = std::tanh(rhs.value());
  const double df = std::pow(1.0f / std::cosh(rhs.value()), 2) * rhs.dot();
  return {value, df};
}

auto coth(const Sym &rhs) noexcept -> Sym {
  const double value = 1.0f / std::tanh(rhs.value());
  const double df = -std::pow(1.0f / std::sinh(rhs.value()), 2) * rhs.dot();
  return {value, df};
}

auto sech(const Sym &rhs) noexcept -> Sym {
  const double value = 1.0f / std::cosh(rhs.value());
  const double df = value * std::tanh(rhs.value()) * rhs.dot();
  return {value, df};
}

auto csch(const Sym &rhs) noexcept -> Sym {
  const double value = 1.0f / std::sinh(rhs.value());
  const double df = -value * (1.0f / std::tanh(rhs.value())) * rhs.dot();
  return {value, df};
}

auto make_jvp() -> std::vector<double>;

}; // namespace ad

/**
 * todo:
 * - higher-order diff in forward mode autodiff
 * - resolve issues with derivative wrt b
 * */

// auto main(void) -> int {
//   auto a = ad::Sym{1.5, 0}; // 0 -- wrt b
//   auto b = ad::Sym{0.5, 1}; // 0 -- wrt a

//   auto f = [](const ad::Sym &a, const ad::Sym &b) -> ad::Sym {
//     return ad::sin(a * b) / ad::cos(b) * ad::pow(a, b);
//   };

//   auto f1 = f(a, b);  // first-order diff wrt b
//   auto f2 = f(f1, b); // second-order diff wrt a

//   std::cout << std::setprecision(15) << f1.dot() << '\n';
//   // std::cout << std::setprecision(15) << f1.value() << '\n';
//   // std::cout << std::setprecision(15) << f2.value() << '\n';
//   std::cout << std::setprecision(15) << f2.dot() << '\n';
// }

File: prototype/py/forward_mode.py
from __future__ import annotations 

from typing import Tuple, Optional, List, Callable
import math as m


sec = lambda x: 1/m.cos(x)

class Symbol:
    def __init__(self, value:float, dot:Optional[float] = None) -> None:
        self.val, self.dot = value, dot

    def __add__(self, other:Symbol) -> Symbol:
        df = Symbol(self.val + other.val,
                self.dot + other.dot)
        return df


    def __add__(self, other:Symbol) -> Symbol:
        df = Symbol(self.val + other.val,
                self.dot + other.dot)
        return df

    def __sub__(self, other:Symbol) -> Symbol:
        df = Symbol(self.val - other.val,
                self.dot - other.dot)
        return df

    def __mul__(self, other:Symbol) -> Symbol:
        df = Symbol(self.val * other.val,
                self.val*other.dot + other.val*self.dot)
        return df

    def __truediv__(self, other:Symbol) -> Symbol:
        df = Symbol(self.val / other.val,
                (self.dot*other.val - self.val*other.dot)/ pow(other.val, 2))
        return df

    def __pow__(self, other:Symbol) -> Symbol:
        df = Symbol(self.val**other.val,
                m.log(self.val)*pow(self.val, other.val)*other.dot)
        return df

def sin(x:Symbol) -> Symbol:
    df = Symbol(m.sin(x.val), m.cos(x.val)*x.dot)
    return df

def cos(x:Symbol) -> Symbol:
    df = Symbol(m.cos(x.val), -m.sin(x.val)*x.dot)
    return df

def tan(x:Symbol) -> Symbol:
    df = Symbol(m.tan(x.val), pow(sec(x.val), 2))
    return df

def make_jvp(fn:List[Callable], 
        args:List[Symbol]) -> List[float]: 
    return [f(*args).dot for f in fn]

if __name__ == '__main__':
    x = Symbol(1.5,0)
    y = Symbol(0.5,1)
    # todo: higher-order diff
    # f = lambda x,y: (x/y - cos(y)) * (x/y - cos(y) + sin(x/y))
    # f1 = f(x,y)
    # f2 = f1/x

    f = [lambda x,y: x*y + sin(x),
         lambda x,y: x+y+sin(x*y)]
    # jvp = make_jvp(f, [x,y])

    print(f[1](x,y).val, f[1](x,y).dot)
